parser-typechecker/src/Unison/Builtin/Decls.hs

```diff
@@ -151,6 +151,86 @@ okConstructorReferent, failConstructorReferent :: Referent.Referent
+unRewriteTerm (Term.Apps' (Term.Constructor' (ConstructorReference r _)) [lhs, rhs])
+  | r == rewriteTermRef = Just (lhs, rhs)
+unRewriteTerm _ = Nothing
+
+rewriteCaseRef :: Reference
+rewriteCaseRef = lookupDeclRef "RewriteCase"
+
+pattern RewriteCase' :: Term2 vt at ap v a -> Term2 vt at ap v a -> Term2 vt at ap v a
+pattern RewriteCase' lhs rhs <- (unRewriteCase -> Just (lhs, rhs))
+
+rewriteCase :: Ord v => a -> Term2 vt at ap v a -> Term2 vt at ap v a -> Term2 vt at ap v a
+rewriteCase a tm1 tm2 = Term.app a (Term.app a1 (Term.constructor a1 r) tm1) tm2
+  where
+    a1 = ABT.annotation tm1
+    r = ConstructorReference rewriteCaseRef 0
+
+rewriteTerm :: Ord v => a -> Term2 vt at ap v a -> Term2 vt at ap v a -> Term2 vt at ap v a
+rewriteTerm a tm1 tm2 = Term.app a (Term.app a1 (Term.constructor a1 r) tm1) tm2
+  where
+    a1 = ABT.annotation tm1
+    r = ConstructorReference rewriteTermRef 0
+
+unRewriteCase :: Term2 vt at ap v a -> Maybe (Term2 vt at ap v a, Term2 vt at ap v a)
+unRewriteCase (Term.Apps' (Term.Constructor' (ConstructorReference r _)) [lhs, rhs])
+  | r == rewriteCaseRef = Just (lhs, rhs)
+unRewriteCase _ = Nothing
+
+rewriteTypeRef :: Reference
+rewriteTypeRef = lookupDeclRef "RewriteSignature"
+
+pattern RewriteSignature' :: forall vt at ap v a. [vt] -> Type vt at -> Type vt at -> Term2 vt at ap v a
+pattern RewriteSignature' vs lhs rhs <- (unRewriteSignature -> Just (vs, lhs, rhs))
+
+rewriteType :: (Var v, Semigroup a) => a -> [v] -> Type v a -> Type v a -> Term2 v a a v a
+rewriteType a vs lhs rhs =
+  Term.app
+    a
+    (Term.constructor la (ConstructorReference rewriteTypeRef 0))
+    ( Term.ann
+        a
+        (Term.delay a (Term.delay a (unitTerm a)))
+        (Type.foralls a vs (Type.arrow (la <> ra) lhs (Type.arrow ra rhs (unitType ra))))
+    )
+  where
+    la = ABT.annotation lhs
+    ra = ABT.annotation rhs
+
+unRewriteSignature :: Term2 vt at ap v a -> Maybe ([vt], Type vt at, Type vt at)
+unRewriteSignature
+  ( Term.App'
+      (Term.Constructor' (ConstructorReference r _))
+      (Term.Ann' _ (Type.ForallsNamedOpt' vs (Type.Arrow' lhs (Type.Arrow' rhs _unit))))
+    )
+    | r == rewriteTypeRef = Just (vs, lhs, rhs)
+unRewriteSignature _ = Nothing
+
+rewritesRef :: Reference
+rewritesRef = lookupDeclRef "Rewrites"
+
+pattern Rewrites' :: [Term2 vt at ap v a] -> Term2 vt at ap v a
+pattern Rewrites' ts <- (unRewrites -> Just ts)
+
+rewrites :: (Var v, Monoid a) => a -> [Term2 vt at ap v a] -> Term2 vt at ap v a
+rewrites a [] = Term.app a (Term.constructor a (ConstructorReference rewritesRef 0)) (tupleTerm [])
+rewrites a ts@(hd : _) = Term.app a (Term.constructor a1 (ConstructorReference rewritesRef 0)) (tupleTerm ts)
+  where
+    a1 = ABT.annotation hd
+
+unRewrites :: Term2 vt at ap v a -> Maybe [Term2 vt at ap v a]
+unRewrites (Term.App' (Term.Constructor' (ConstructorReference r _)) tup)
+  | r == rewritesRef, TupleTerm' ts <- tup = Just ts
+unRewrites _ = Nothing
```

Comment:
`@rewrite` blocks are encoded as values of type `Rewrites a`, where `a` is a tuple of `RewriteTerm`, `RewriteCase`, and/or `RewriteSignature` values.